File: app/src/main/java/com/example/win/easy/repository/db/dao/SongDao.java
package com.example.win.easy.repository.db.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import com.example.win.easy.repository.db.data_object.SongDO;

import java.util.Collection;
import java.util.List;

@Dao
public interface SongDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insert(SongDO songDO);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long[] insert(Collection<SongDO> songDOs);

    //以下为Repo涉及方法
    @Query("SELECT * FROM SongDO")
    List<SongDO> getAllSong();

    @Query("SELECT * FROM SongDO WHERE remoteId=:RemoteId")
    SongDO findByRemoteId(Long RemoteId);

    @Query("SELECT * FROM SongDO WHERE remoteId is not null")
    List<SongDO> findAllDataOnWeb();

    @Query("SELECT * FROM SongDO WHERE remoteId is null")
    List<SongDO> findAllDataOnLocal();
    /***********************/

    @Query("SELECT * FROM SongDO")
    LiveData<List<SongDO>> findAllSongDOs();

    @Query("SELECT * FROM SongDO WHERE id=:songId")
//    SongDO findById(Long );songId
    LiveData<SongDO> findById(Long songId);

    @Query("SELECT * FROM SongDO WHERE id=:songId")
    SongDO findDataById(Long songId);

    @Query("SELECT * FROM SongDO WHERE source=:dataSource")
//    List<SongDO> findByDataSource(String dataSource);
    LiveData<List<SongDO>> findByDataSource(String dataSource);

    @Query("SELECT * FROM SongDO WHERE source!=:dataSource")
    List<SongDO> findAllSongsExceptByDataSource(String dataSource);

    @Query("SELECT * FROM SongDO WHERE songPath=:songPath")
    SongDO findAllBySongPath(String songPath);

    @Delete
    void delete(SongDO songDO);

    @Delete
    void delete(Collection<SongDO> songDOs);

    @Update
    void update(SongDO songDO);

    @Update
    void update(Collection<SongDO> songDOs);
}

File: app/src/main/java/com/example/win/easy/repository/db/dao/SongDao.kt
package com.example.win.easy.repository.db.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import com.example.win.easy.repository.db.data_object.SongDO

@Dao
interface SongDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(songDO: SongDO): Long

    @Query("SELECT * FROM SongDO WHERE songPath is not null")
    fun loadAllLocal(): LiveData<List<SongDO>?>

    @Query("SELECT * FROM SongDO WHERE songPath=:path")
    suspend fun findByPath(path: String): List<SongDO>

    @Update
    suspend fun update(songDO: SongDO)

    //以下为Repo涉及方法
    @Query("SELECT * FROM SongDO")
    fun findAllSong(): List<SongDO>

    @Query("SELECT * FROM SongDO WHERE remoteId=:RemoteId")
    fun findByRemoteId(RemoteId: Long?): SongDO?

    @Query("SELECT * FROM SongDO WHERE remoteId is not null")
    fun findAllDataOnWeb(): List<SongDO?>?

    @Query("SELECT * FROM SongDO WHERE remoteId is null")
    fun findAllDataOnLocal(): List<SongDO?>?


    /** */
    @Query("SELECT * FROM SongDO")
    fun findAllSongDOs(): LiveData<List<SongDO?>?>?

    @Query("SELECT * FROM SongDO WHERE songPath=:songPath")
    fun findAllBySongPath(songPath: String?): SongDO?

    @Delete
    fun delete(songDO: SongDO?)


}